source/uwp/Renderer/lib/XamlBuilder.cpp

```diff
@@ -2654,10 +2655,202 @@ AdaptiveNamespaceStart
+            THROW_IF_FAILED(titleTextBlock.As(&textBlockAsFrameworkElement));
+            THROW_IF_FAILED(textBlockAsFrameworkElement->put_VerticalAlignment(VerticalAlignment_Center));
+
+            ComPtr<IGrid> titleGrid = XamlHelpers::CreateXamlClass<IGrid>(HStringReference(RuntimeClass_Windows_UI_Xaml_Controls_Grid));
+            ComPtr<IPanel> panel;
+            THROW_IF_FAILED(titleGrid.As(&panel));
+            XamlHelpers::AppendXamlElementToPanel(titleTextBlock.Get(), panel.Get());
+
+            THROW_IF_FAILED(panel.As(&actionUIElement));
+        }
+
+        // Make the action the same size as the text box
+        EventRegistrationToken eventToken;
+        THROW_IF_FAILED(textBoxAsFrameworkElement->add_Loaded(Callback<IRoutedEventHandler>(
+            [actionUIElement, textBoxAsFrameworkElement](IInspectable* /*sender*/, IRoutedEventArgs* /*args*/) -> HRESULT
+        {
+            ComPtr<IFrameworkElement> actionFrameworkElement;
+            RETURN_IF_FAILED(actionUIElement.As(&actionFrameworkElement));
+
+            return SetMatchingHeight(actionFrameworkElement.Get(), textBoxAsFrameworkElement.Get());
+        }).Get(), &eventToken));
+
+        // Wrap the action in a button
+        ComPtr<IUIElement> touchTargetUIElement;
+        WrapInTouchTarget(
+            nullptr, 
+            actionUIElement.Get(), 
+            localInlineAction.Get(), 
+            renderContext, 
+            false, 
+            L"Adaptive.Input.Text.InlineAction",
+            &touchTargetUIElement);
+
+        ComPtr<IFrameworkElement> touchTargetFrameworkElement;
+        THROW_IF_FAILED(touchTargetUIElement.As(&touchTargetFrameworkElement));
+
+        // Align to bottom so the icon stays with the bottom of the text box as it grows in the multiline case
+        THROW_IF_FAILED(touchTargetFrameworkElement->put_VerticalAlignment(VerticalAlignment_Bottom));
+
+        // Add the action to the column
+        THROW_IF_FAILED(gridStatics->SetColumn(touchTargetFrameworkElement.Get(), 2));
+        XamlHelpers::AppendXamlElementToPanel(touchTargetFrameworkElement.Get(), gridAsPanel.Get());
+
+        // If this isn't a multiline input, enter should invoke the action
+        ComPtr<IAdaptiveActionInvoker> actionInvoker;
+        THROW_IF_FAILED(renderContext->get_ActionInvoker(&actionInvoker));
+
+        boolean isMultiLine;
+        THROW_IF_FAILED(textBox->get_AcceptsReturn(&isMultiLine));
+
+        if (!isMultiLine)
+        {
+            EventRegistrationToken keyDownEventToken;
+            THROW_IF_FAILED(textBoxAsUIElement->add_KeyDown(Callback<IKeyEventHandler>(
+                [touchTargetUIElement, actionInvoker, localInlineAction](IInspectable* /*sender*/, IKeyRoutedEventArgs* args) -> HRESULT
+            {
+                ABI::Windows::System::VirtualKey key;
+                RETURN_IF_FAILED(args->get_Key(&key));
+
+                if (key == ABI::Windows::System::VirtualKey::VirtualKey_Enter)
```

Comment:
>if (key == ABI::Windows::System::VirtualKey::VirtualKey_Enter) [](start = 16, length = 62)

more of a spec question -- do we care about shift-enter?

Reply:
So, in fact, the spec calls for multiline text boxes to submit on enter and add a newline on shift-enter, but only if the user is using a physical keyboard. I ran into a few issues with that, and I've got a mail out to Andrew to clarify, so when he returns there may be more changes here in another PR. The behavior implemented in this PR is that in multiline text boxes enter is a newline and does not trigger the action (which is the spec'd behavior for touch keyboards).

That said, even in this case, which is a non-multiline text box, we shouldn't submit on shift-enter. I'll update this to not invoke the action if shift is pressed.

---
In reply to: [223483878](https://github.com/Microsoft/AdaptiveCards/pull/1997#discussion_r223483878) [](ancestors = 223483878)